### PR TITLE
Add possibility to do nested cross validation

### PR DESCRIPTION
**Pull request by PhilipMay**
This adds the possibility to do nested cross validation.

The PR introduces a new class `DataSiloForNestedCrossVal` that inherits from `DataSiloForCrossVal`.
It does nested cross validation:

- other cross validation splits the data to a rest set and a test set
- inner cross validation splits the rest data to train and dev

This is related to #507

## Todo
- [x] write test
- [x] discuss with @Timoeller (and @johann-petrak) 
- [ ] ~~add quest. answering~~
- [x] add doc
- [x] add number of folds pram for outer and inner folds

### CODE REVIEW

**Comment by Timoeller** [comment]
After some offline discussion with @PhilipMay we are going forward with the nested Crossvalidation. 

We need a way to handle the variance in performance that comes from different restarts (be it different seeds or slightly different training data). 

This PR adds a solution in a structured way, so glad to have it in FARM.

**Comment by PhilipMay** [comment]
An other plus for this nested cross validation is that the train and dev datasets are also created in a stratified way.

**Comment by PhilipMay** [comment]
Rebased on master.

File: test/data_handler/test_data_silo.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
+from pathlib import Path
```
Comment (Timoeller) [inline test/data_handler/test_data_silo.py:1]: lets leave this fct in the top level folder with all the other tests.
If we restructure the test folder we will do so in a separate PR and for all tests

File: test/data_handler/test_data_silo.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+import numpy as np
+
+from farm.modeling.tokenization import Tokenizer
+from farm.data_handler.processor import TextClassificationProcessor
+from farm.data_handler.data_silo import DataSilo, DataSiloForNestedCrossVal
+
+
+def test_data_silo_for_nested_cross_val():
+    lang_model = "bert-base-german-cased"
+    n_outer_splits = 7
```
Comment (Timoeller) [inline test/data_handler/test_data_silo.py:11]: lets use less splits, otherwise processing takes longer than necessary. Or is there a special reason for these numbers?

File: farm/data_handler/data_silo.py [hunk 3]
```diff
@@ -942,3 +942,104 @@ def _split_for_qa(documents, id_index, n_splits=5, shuffle=True, random_state=No
...
+        :param random_state: random state for shuffling
+        :type random_state: int
+        :param stratified: if class stratification should be done
+        :type stratified: bool
+        :param n_neg_answers_per_question: number of negative answers per question to include for training
+        :type n_neg_answers_per_question: int
+        """
+
+        if "question_answering" in datasilo.processor.tasks:
+            return cls._make_question_answering(datasilo, sets, n_splits, shuffle, random_state, n_neg_answers_per_question)
```
Comment (Timoeller) [inline farm/data_handler/data_silo.py:996]: n_splits is not defined and will throw an error. We should nevertheless try to integrate the inner split functionality into DataSiloForCrossVal

File: farm/data_handler/data_silo.py [hunk 4]
```diff
@@ -942,3 +942,104 @@ def _split_for_qa(documents, id_index, n_splits=5, shuffle=True, random_state=No
             current_train_set = np.hstack(np.delete(splits, idx, axis=0))
 
             yield current_train_set, current_test_set
+
+
+class DataSiloForNestedCrossVal(DataSiloForCrossVal):
```
Comment (Timoeller) [inline farm/data_handler/data_silo.py:947]: Now that I see the code there is a lot of duplicate code in this class and DataSiloForCrossVal.
Cant we merge both classes and parameterize it?

File: farm/data_handler/data_silo.py [hunk 5]
```diff
@@ -942,3 +942,104 @@ def _split_for_qa(documents, id_index, n_splits=5, shuffle=True, random_state=No
...
+    """
+    Prepare data silos to do a nested cross validation.
+
+    The outer cross validation splits all data to test and rest.
+    The inner cross validation splits the rest data to train and dev.
```
Comment (Timoeller) [inline farm/data_handler/data_silo.py:952]: Lets motive the inner split in the docstring by talking about variance in multiple restarts

**Review by Timoeller — CHANGES_REQUESTED** [review]
This is looking well but could be a bit better integrated into existing code.
I made some comments along the way - open to discussion

File: test/data_handler/test_data_silo.py (hunk 1 above)
Comment (PhilipMay) [reply test/data_handler/test_data_silo.py]: done.

File: test/data_handler/test_data_silo.py (hunk 2 above)
Comment (PhilipMay) [reply test/data_handler/test_data_silo.py]: yes - done

File: farm/data_handler/data_silo.py (hunk 3 above)
Comment (PhilipMay) [reply farm/data_handler/data_silo.py]: fixed

File: farm/data_handler/data_silo.py (hunk 4 above)
Comment (PhilipMay) [reply farm/data_handler/data_silo.py]: I have done that. @Timoeller what do you think about this solution?

File: farm/data_handler/data_silo.py (hunk 5 above)
Comment (PhilipMay) [reply farm/data_handler/data_silo.py]: I added extensive documentation. :-)

**Comment by PhilipMay** [comment]
@Timoeller I made your suggested changes.
I hope the integration into the existing class is fine with you.

I had a look on nested cross validation for question answering.
But to be honest the whole thing is a bit too fiddly for me. 
I'm not deep enough in the question answering topic and would build many bugs I guess.

Maybe you could just add this later when you / someone else wants / needs it?

IMO this is good to go to be merged.

**Review by Timoeller — APPROVED** [review]
Looking good, thanks for integrating the proposed changes.
